drivers/storage/ide/atapi/dev_identify.c

```diff
@@ -0,0 +1,359 @@
+    _In_ PIDENTIFY_DEVICE_DATA IdentifyData2)
+{
+    if (!RtlEqualMemory(IdentifyData1->SerialNumber,
+                        IdentifyData2->SerialNumber,
+                        RTL_FIELD_SIZE(IDENTIFY_DEVICE_DATA, SerialNumber)))
+    {
+        return FALSE;
+    }
+
+    if (!RtlEqualMemory(IdentifyData1->FirmwareRevision,
+                        IdentifyData2->FirmwareRevision,
+                        RTL_FIELD_SIZE(IDENTIFY_DEVICE_DATA, FirmwareRevision) +
+                        RTL_FIELD_SIZE(IDENTIFY_DEVICE_DATA, ModelNumber)))
+    {
+        return FALSE;
+    }
+
+    return TRUE;
+}
+
+NTSTATUS
+AtaDeviceSendIdentify(
+    _In_ PATAPORT_PORT_DATA PortData,
+    _In_ PATAPORT_DEVICE_EXTENSION DevExt,
+    _In_ UCHAR Command)
+{
+    PATA_DEVICE_REQUEST Request = &PortData->Worker.InternalRequest;
+
+    TRACE("CH %lu: Send %s identify device %u\n",
+          PortData->PortNumber,
+          Command == IDE_COMMAND_IDENTIFY ? "ATA" : "ATAPI",
+          DevExt->Device.AtaScsiAddress.TargetId);
+
+    /*
+     * For PATA devices, disable interrupts for the identify command and use polling instead.
+     *
+     * On some single device 1 configurations or non-existent IDE channels
+     * the status register stuck permanently at a value of 0,
+     * and we incorrectly assume that the device is present.
+     * This will result in a taskfile timeout
+     * which must be avoided as it would cause hangs at boot time.
+     */
+    Request->Flags = REQUEST_FLAG_POLL;
+
+    Request->Flags |= REQUEST_FLAG_DATA_IN | REQUEST_FLAG_HAS_LOCAL_BUFFER;
+    Request->TimeOut = 10;
+    Request->DataTransferLength = sizeof(DevExt->IdentifyDeviceData);
+
+    RtlZeroMemory(&Request->TaskFile, sizeof(Request->TaskFile));
+    Request->TaskFile.Command = Command;
+
+    return AtaPortSendRequest(PortData, DevExt);
+}
+
+static
+NTSTATUS
+AtaDeviceSpinUp(
+    _In_ PATAPORT_PORT_DATA PortData,
+    _In_ PATAPORT_DEVICE_EXTENSION DevExt)
+{
+    PATA_DEVICE_REQUEST Request = &PortData->Worker.InternalRequest;
+
+    TRACE("CH %lu: Spin-up device %u\n",
+          PortData->PortNumber,
+          DevExt->Device.AtaScsiAddress.TargetId);
+
+    Request->Flags = 0;
+    Request->TimeOut = 20;
+
+    Request->TaskFile.Command = IDE_COMMAND_SET_FEATURE;
+    Request->TaskFile.Feature = IDE_FEATURE_PUIS_SPIN_UP;
+
+    return AtaPortSendRequest(PortData, DevExt);
+}
+
+static
+BOOLEAN
+AtaDeviceIsXboxDrive(
+    _In_ PATAPORT_DEVICE_EXTENSION DevExt)
+{
+    static const PCSTR AtapBrokenInquiryDrive[] =
+    {
+        "THOMSON-DVD",
+        "PHILIPS XBOX DVD DRIVE",
+        "PHILIPS J5 3235C",
+        "SAMSUNG DVD-ROM SDG-605B"
+        // TODO: Xbox 360 drives are also affected.
+    };
+    CHAR ModelNumber[26];
+    PUCHAR End;
+    ULONG i;
+
+    End = AtaCopyIdStringUnsafe((PUCHAR)ModelNumber,
+                                DevExt->IdentifyDeviceData.ModelNumber,
+                                sizeof(ModelNumber));
+    *End = ANSI_NULL;
```

Comment:
This looks like a buffer overrun. If the goal is to terminate the buffer, this should probably be done by `AtaCopyIdStringUnsafe` or do it here as `ModelNumber[25] = ANSI_NULL`.
And should this is really be `ANSI_NULL`, when `AtaCopyIdStringUnsafe` shuffles around `USHORT`s?

Reply:
Shoud be fixed now.